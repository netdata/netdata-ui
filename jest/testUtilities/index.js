import "@testing-library/jest-dom/extend-expect"

export { default as renderWithProviders } from "./renderWithProviders"
export { default as renderHookWithProviders } from "./renderHookWithProviders"

import userEvent from "@testing-library/user-event"
export const user = () => userEvent.setup()
export { userEvent }

export * from "@testing-library/react"
