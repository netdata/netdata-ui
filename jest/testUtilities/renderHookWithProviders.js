import React from "react"
import { renderHook } from "."
import { ThemeProvider } from "styled-components"
import { DefaultTheme } from "@/theme/default"

export default (hook, { theme = DefaultTheme, ...rest } = {}) =>
  renderHook(hook, {
    wrapper: props => <ThemeProvider theme={theme} {...props} />,
    ...rest,
  })
