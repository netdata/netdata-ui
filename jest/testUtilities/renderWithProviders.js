import React from "react"
import { render } from "."
import { ThemeProvider } from "styled-components"
import { DefaultTheme } from "@/theme/default"

export default (Component, { theme = DefaultTheme, ...rest } = {}) =>
  render(Component, {
    wrapper: props => <ThemeProvider theme={theme} {...props} />,
    ...rest,
  })
