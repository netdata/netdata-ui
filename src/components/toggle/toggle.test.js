/**
 * @jest-environment jsdom
 */

import React, { useState } from "react"
import { renderWithProviders, fireEvent } from "testUtilities"
import { Toggle } from "."

const MockedToggle = props => {
  const [checked, setChecked] = useState(false)
  const handleChange = e => {
    setChecked(e.currentTarget.checked)
  }
  return <Toggle onChange={handleChange} checked={checked} {...props} />
}

describe("Toggle test", () => {
  it(" * should render with required props", () => {
    const { container, getByRole } = renderWithProviders(<MockedToggle />)
    const checkbox = container.querySelectorAll("input")[0]
    expect(checkbox).toBeInTheDocument()

    const toggle = getByRole("switch")
    expect(toggle).toHaveStyleRule("background", "#FFFFFF")
    expect(toggle).toHaveStyleRule("background-color", "#35414A", { modifier: ":after" })
    expect(container).toMatchSnapshot()
  })

  it(" * should render with labels", () => {
    const { container, getByText } = renderWithProviders(
      <MockedToggle labelRight="Light theme" labelLeft="Dark theme" />
    )
    expect(getByText(/Light theme/)).toBeInTheDocument()
    expect(getByText(/Dark theme/)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it(" * should be disabled", () => {
    const { container, getByRole } = renderWithProviders(<MockedToggle disabled />)
    const checkbox = container.querySelectorAll("input")[0]
    expect(checkbox).toBeDisabled()

    const toggle = getByRole("switch")
    expect(toggle).toHaveStyleRule("background-color", "#8F9EAA", { modifier: ":after" })
  })

  it(" * should be colored", () => {
    const { container, getByRole } = renderWithProviders(<MockedToggle colored />)
    const toggle = getByRole("switch")
    expect(toggle).toHaveStyleRule("background-color", "#FF4136", { modifier: ":after" })

    const checkbox = container.querySelectorAll("input")[0]
    fireEvent.click(checkbox)
    expect(toggle).toHaveStyleRule("background-color", "#00AB44", { modifier: ":after" })
  })

  it(" * should check the checkbox on interaction", () => {
    const { container } = renderWithProviders(<MockedToggle />)
    const checkbox = container.querySelectorAll("input")[0]
    fireEvent.click(checkbox)
    expect(checkbox.checked).toBe(true)

    fireEvent.click(checkbox)
    expect(checkbox.checked).toBe(false)
  })
})
