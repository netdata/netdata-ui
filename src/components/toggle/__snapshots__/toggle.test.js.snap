// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle test  * should render with labels 1`] = `
.c1 {
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Ubuntu,"Helvetica Neue",sans-serif;
  font-style: normal;
  font-size: 14px;
  line-height: 20px;
  font-weight: normal;
  color: #536775;
  text-transform: none;
}

.c3 {
  display: block;
  box-sizing: border-box;
  width: 40px;
  height: 20px;
}

.c5 {
  display: none;
}

.c7 {
  box-sizing: border-box;
  width: 40px;
  height: 20px;
  background: #FFFFFF;
  border: 1px solid #8F9EAA;
  border-radius: 100px;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  display: block;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  cursor: pointer;
  pointer-events: auto;
}

.c7:after {
  display: block;
  position: absolute;
  content: "";
  width: 16px;
  height: 16px;
  border-radius: 50%;
  left: 5%;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: left 0.2s ease;
  transition: left 0.2s ease;
  background-color: #35414A;
}

.c4:focus + .c6 {
  border-color: #35414A;
  box-shadow: 0 0 0 1px #35414A;
}

.c0 {
  position: relative;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  margin-right: 8px;
}

.c8 {
  margin-left: 8px;
}

<div>
  <label
    class="c0"
  >
    <span
      class="c1 c2"
    >
      Dark theme
    </span>
    <div
      class="c3"
    >
      <input
        class="c4 c5"
        type="checkbox"
      />
      <div
        class="c6 c7"
        role="switch"
      />
    </div>
    <span
      class="c1 c8"
    >
      Light theme
    </span>
  </label>
</div>
`;

exports[`Toggle test  * should render with required props 1`] = `
.c1 {
  display: block;
  box-sizing: border-box;
  width: 40px;
  height: 20px;
}

.c3 {
  display: none;
}

.c5 {
  box-sizing: border-box;
  width: 40px;
  height: 20px;
  background: #FFFFFF;
  border: 1px solid #8F9EAA;
  border-radius: 100px;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  display: block;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  cursor: pointer;
  pointer-events: auto;
}

.c5:after {
  display: block;
  position: absolute;
  content: "";
  width: 16px;
  height: 16px;
  border-radius: 50%;
  left: 5%;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: left 0.2s ease;
  transition: left 0.2s ease;
  background-color: #35414A;
}

.c2:focus + .c4 {
  border-color: #35414A;
  box-shadow: 0 0 0 1px #35414A;
}

.c0 {
  position: relative;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <label
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2 c3"
        type="checkbox"
      />
      <div
        class="c4 c5"
        role="switch"
      />
    </div>
  </label>
</div>
`;
